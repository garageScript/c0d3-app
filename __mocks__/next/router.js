const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
