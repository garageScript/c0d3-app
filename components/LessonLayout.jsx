import React from 'react'
import LessonHeader from './LessonHeader'
import Layout from './Layout'

const LessonLayout = ({ children, ...rest }) => (
  <>
    <LessonHeader {...rest} />
    {children}
  </>
)

export const getLayout = (page, { metaData }) => {
  return (
    <Layout>
      <LessonLayout metaData={metaData} isPassed={true}>
        {page}
      </LessonLayout>
    </Layout>
  )
}

export default LessonLayout
