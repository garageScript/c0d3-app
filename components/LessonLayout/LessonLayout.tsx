import React, { useEffect, useState } from 'react'
import Layout from '../Layout'
import LessonTitleCard from '../LessonTitleCard'

import { LayoutGetter } from '../../@types/page'
import { Lesson, useGetSessionQuery } from '../../graphql'

type Props = {
  lesson: Lesson
}
const LessonLayout: React.FC<Props> = ({ children, lesson }) => {
  const [isPassed, setIsPassed] = useState(false)
  const { data: { session } = {} } = useGetSessionQuery({ ssr: false })
  useEffect(() => {
    if (!session) {
      setIsPassed(false)
      return
    }
    const newState = Boolean(
      session.lessonStatus?.find(l => l.lessonId === lesson.id)?.passedAt
    )

    setIsPassed(newState)
  }, [session, lesson.id])
  return (
    <div className="row mt-4">
      <LessonTitleCard
        lessonCoverUrl={`js-${lesson.order}-cover.svg`}
        lessonUrl={lesson.docUrl}
        lessonTitle={lesson.title}
        lessonId={lesson.id}
        lessonSlug={lesson.slug}
        isPassed={isPassed}
      />
      {children}
    </div>
  )
}

export const getLayout: LayoutGetter<Props> = (page, { lesson }) => {
  return (
    <Layout>
      {/* TODO: Add fallback page title */}
      <LessonLayout lesson={lesson}>{page}</LessonLayout>
    </Layout>
  )
}

export default LessonLayout
