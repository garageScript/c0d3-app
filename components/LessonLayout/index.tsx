export { default, getLayout } from './LessonLayout'
