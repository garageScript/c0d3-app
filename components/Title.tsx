import React from 'react'
import Head from 'next/head'

const Title: React.FC<{ title?: string }> = ({ title }) => (
  <Head>
    <title>{title ? `${title} ? C0D3` : 'C0D3'}</title>
  </Head>
)

export default Title
