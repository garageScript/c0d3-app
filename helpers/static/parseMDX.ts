import { serialize } from 'next-mdx-remote/serialize'
import { MDXRemoteSerializeResult } from 'next-mdx-remote'
import matter from 'gray-matter'
import gfm from 'remark-gfm'
import slug from 'remark-slug'
import autolink from 'rehype-autolink-headings'

type ParsedMDX = {
  (
    fileContents: Buffer,
    onlyFront?: boolean
  ): Promise<{
    source?: MDXRemoteSerializeResult<Record<string, unknown>>
    frontMatter: { [key: string]: any }
  }>
}

const getSource = (content: string) =>
  serialize(content, {
    mdxOptions: {
      remarkPlugins: [slug, gfm],
      rehypePlugins: [autolink]
    }
  })

export const parseMDX: ParsedMDX = async (fileContents, onlyFront = false) => {
  const { content, data } = matter(fileContents)

  if (onlyFront) return { frontMatter: data }

  return {
    frontMatter: data,
    source: await getSource(content)
  }
}
