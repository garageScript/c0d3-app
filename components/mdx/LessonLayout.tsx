import React, { useEffect, useState } from 'react'
import Layout from '../Layout'
import LessonTitleCard, { LessonTitleProps } from '../LessonTitleCard'
import styles from '../../scss/mdx.module.scss'
import { useRouter } from 'next/router'
import _ from 'lodash'

type LayoutProps = Partial<
  LessonTitleProps & { subLessons?: string[]; title?: string }
>

const ScrollTop: React.FC<{ scroll: boolean }> = ({ scroll }) => {
  return (
    <img
      src="/assets/mdx/topArrow.svg"
      className={`${scroll ? 'd-block' : 'd-none'} position-fixed ${
        styles['arrow']
      }`}
      onClick={() =>
        window.scrollTo({
          top: 0,
          left: 0,
          behavior: 'smooth'
        })
      }
      data-testid="arrow"
    />
  )
}
const LessonLayout: React.FC<LayoutProps> = props => {
  const router = useRouter()
  const location = router.pathname.split('/')
  const current = location[location.length - 1]
  const [scroll, setScroll] = useState(false)
  useEffect(() => {
    const throttled = _.throttle(() => {
      setScroll(
        window.scrollY > 2 * window.document.documentElement.clientHeight
      )
    }, 100)
    window.addEventListener('scroll', throttled)
  }, [])
  const lessonParts =
    props.subLessons &&
    props.subLessons.map((l, i) => {
      const url = l.replace(',', '').toLowerCase().split(' ').join('_')
      return (
        <a
          className={`${styles['subtitle']} ${
            url !== current ? 'text-muted' : `text-dark dark`
          } d-block`}
          href={url}
          key={url}
        >
          {`Part ${i + 1}: ${l}`}
        </a>
      )
    })
  return (
    <Layout title={props.title || 'C0D3'}>
      <div className="mt-4">
        {props.lessonCoverUrl && (
          <LessonTitleCard
            {...(props as LessonTitleProps)}
            lessonTitle={props.title!}
          />
        )}
      </div>
      <div
        className={`${styles['lesson-wrapper']} card shadow-sm mt-3 d-block border-0 p-3 p-md-4 bg-white`}
      >
        {!props.lessonCoverUrl && (
          <button
            className="btn btn-link text-primary p-0"
            onClick={() => router.back()}
          >
            Go Back
          </button>
        )}
        <ScrollTop scroll={scroll} />
        <div className={styles.title}>{props.title}</div>
        {lessonParts}
        {props.children}
      </div>
    </Layout>
  )
}
export default LessonLayout
