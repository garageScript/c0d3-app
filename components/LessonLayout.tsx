import React, { useEffect, useState } from 'react'
import Layout from './Layout'
import LessonTitleCard, { LessonTitleProps } from './LessonTitleCard'
import styles from '../scss/mdx.module.scss'
import { useRouter } from 'next/router'
import _ from 'lodash'
import { LayoutGetter } from '../@types/page'
import { LessonMetaData } from '../helpers/static/getLessonMetaData'
type LayoutProps = {
  metaData: LessonMetaData
}

const ScrollTopArrow: React.FC<{ scroll: boolean }> = ({ scroll }) => {
  return (
    <img
      src="/assets/mdx/topArrow.svg"
      className={`${scroll ? 'd-block' : 'd-none'} position-fixed ${
        styles['arrow']
      }`}
      onClick={() =>
        window.scrollTo({
          top: 0,
          left: 0,
          behavior: 'smooth'
        })
      }
      data-testid="arrow"
    />
  )
}

const EditPage: React.FC = () => {
  const router = useRouter()
  return (
    <a
      href={`https://github.com/garageScript/c0d3-app/edit/master/pages/${router.pathname}.mdx`}
      className={`${styles['MDX_a']} ${styles['edit']}`}
    >
      Edit this page
    </a>
  )
}
const LessonLayout: React.FC<LayoutProps> = ({ children, metaData }) => {
  const router = useRouter()
  const [scroll, setScroll] = useState(false)
  useEffect(() => {
    const throttled = _.throttle(() => {
      setScroll(
        window.scrollY > 2 * window.document.documentElement.clientHeight
      )
    }, 100)
    window.addEventListener('scroll', throttled)
  }, [])

  return (
    <>
      <input type="text"></input>
      <LessonTitleCard metaData={metaData} isPassed={true} />

      <ScrollTopArrow scroll={scroll} />

      {children}
    </>
  )
}
export const getLayout: LayoutGetter = (page, { metaData }) => {
  return (
    <Layout>
      {/* TODO: Add fallback page title */}
      <LessonLayout metaData={metaData}>{page}</LessonLayout>
    </Layout>
  )
}

export default LessonLayout
