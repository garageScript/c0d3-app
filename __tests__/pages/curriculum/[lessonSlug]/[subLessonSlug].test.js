jest.mock('gray-matter')
import matter from 'gray-matter'
import { mockUseBreakpoint } from '../../../../__mocks__/useBreakpoint.mock'
import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import SubLessonPage, {
  getStaticPaths,
  getStaticProps
} from '../../../../pages/curriculum/[lessonSlug]/[subLessonSlug]'
import {
  dummyParsedSubLessonMdx,
  dummySubLessonFileContent
} from '../../../../__dummy__/mdx'
import { useRouter } from 'next/router' // Auto mocked
import { getLayout } from '../../../../components/LessonLayout'
jest.mock('../../../../helpers/apolloClient-server')
import { initializeApollo } from '../../../../helpers/apolloClient-server'
import Title from '../../../../components/Title'
jest.mock('../../../../components/Title', () => {
  return jest.fn(() => null)
})
import {
  getSubLessonSlugs,
  getSubLessonContent,
  getSubLessonGithubFilePath
} from '../../../../helpers/static/lessons'
import { parseMDX } from '../../../../helpers/static/parseMDX'
jest.mock('../../../../helpers/static/parseMDX')
import dummyLessonsData from '../../../../__dummy__/lessonData'
jest.mock('../../../../helpers/static/lessons')
import userEvent from '@testing-library/user-event'

describe('[subLessonSlug]', () => {
  const mockSlugs = [
    { lessonSlug: 'js0', subLessonSlug: 'first_sub_lesson' },
    { lessonSlug: 'js0', subLessonSlug: 'second_sub_lesson' },
    { lessonSlug: 'js0', subLessonSlug: 'third_sub_lesson' }
  ]
  const mockHeadings = [
    { text: 'heading1', depth: 1 },
    { text: 'heading2', depth: 2 },
    { text: 'heading3', depth: 3 }
  ]
  const fakeGithubPath = 'some/fake/path'

  const dummyOneFrontMatterOnly = {
    frontMatter: dummyParsedSubLessonMdx[1].frontMatter
  }
  const dummyTwoFrontMatterOnly = {
    frontMatter: dummyParsedSubLessonMdx[2].frontMatter
  }
  const { challenges, ...dummyLessonNoChallenges } = dummyLessonsData[0] // Lesson with slug 'js0'
  const props = includeHeadings => ({
    lesson: dummyLessonNoChallenges,
    lessonSlug: mockSlugs[0].lessonSlug,

    selectedSubLessonIndex: 0,
    subLessonSlug: mockSlugs[0].subLessonSlug,
    subLessons: [
      {
        ...dummyParsedSubLessonMdx[0],
        subLessonSlug: mockSlugs[0].subLessonSlug,
        headings: includeHeadings ? [] : mockHeadings
      },
      {
        ...dummyOneFrontMatterOnly,
        subLessonSlug: mockSlugs[1].subLessonSlug,
        headings: includeHeadings ? [] : mockHeadings
      },
      {
        ...dummyTwoFrontMatterOnly,
        subLessonSlug: mockSlugs[2].subLessonSlug,
        headings: includeHeadings ? [] : mockHeadings
      }
    ],

    githubFilePath: fakeGithubPath
  })

  beforeEach(() => {
    jest.clearAllMocks()
    matter.mockReturnValue({
      content: `
# heading1
## heading2
### heading3
        `
    })
  })

  describe('getStaticPaths', () => {
    test('should return paths w/ fallback false', () => {
      getSubLessonSlugs.mockResolvedValue(mockSlugs)

      expect(getStaticPaths()).resolves.toEqual({
        paths: [
          { params: mockSlugs[0] },
          { params: mockSlugs[1] },
          { params: mockSlugs[2] }
        ],
        fallback: false
      })
    })

    test('should throw if invalid slug names are retrieved', () => {
      getSubLessonSlugs.mockResolvedValue([
        { old_slug_name: 'js0', another_bad_name: 'first_sub_lesson' }
      ])
      expect(getStaticPaths()).rejects.toThrowError()
    })
  })

  describe('getStaticProps', () => {
    test('should throw when called with no slug', () => {
      expect(getStaticProps({ params: {} })).rejects.toThrowError(
        /Missing Slug/
      )
    })

    test('should throw if it can not find matching lesson', () => {
      const mockQuery = jest.fn().mockResolvedValue({ data: { lessons: [] } })
      initializeApollo.mockReturnValue({ query: mockQuery })
      expect(
        getStaticProps({
          params: { lessonSlug: 'js100', subLessonSlug: 'no_way' }
        })
      ).rejects.toThrowError(/Could not find lesson/)
    })

    test('should return correct props', () => {
      const mockQuery = jest
        .fn()
        .mockResolvedValue({ data: { lessons: dummyLessonsData } })
      initializeApollo.mockReturnValue({ query: mockQuery })
      getSubLessonGithubFilePath.mockReturnValue(fakeGithubPath)
      getSubLessonSlugs.mockResolvedValueOnce(mockSlugs)
      getSubLessonContent
        .mockResolvedValueOnce(dummySubLessonFileContent[0])
        .mockResolvedValueOnce(dummySubLessonFileContent[1])
        .mockResolvedValueOnce(dummySubLessonFileContent[2])

      parseMDX
        .mockResolvedValueOnce(dummyParsedSubLessonMdx[0])
        .mockResolvedValueOnce(dummyOneFrontMatterOnly)
        .mockResolvedValueOnce(dummyTwoFrontMatterOnly)

      expect(getStaticProps({ params: mockSlugs[0] })).resolves.toEqual({
        props: props(false),
        revalidate: 300 // Five Minutes
      })
    })
  })

  describe('Page', () => {
    test('Should have static method getLayout from LessonLayout component', async () => {
      expect(SubLessonPage.getLayout).toBe(getLayout)
    })

    test('should have correct title', () => {
      render(<SubLessonPage {...props(false)} />)

      expect(Title).toHaveBeenCalledWith(
        {
          title: expect.stringContaining('first sub lesson')
        },
        {}
      )
    })

    test('should render SubLessonLinks component with correct subLesson titles', () => {
      render(<SubLessonPage {...props(false)} />)

      expect(
        screen.getByRole('link', { name: 'Part 1: first sub lesson' })
      ).toBeVisible()
      expect(
        screen.getByRole('link', { name: 'Part 2: second sub lesson' })
      ).toBeVisible()
      expect(
        screen.getByRole('link', { name: 'Part 3: third sub lesson' })
      ).toBeVisible()
    })
    test('should render next lesson link', () => {
      render(<SubLessonPage {...props(false)} />)

      expect(
        screen.getByRole('link', { name: 'Next part: second sub lesson' })
      ).toBeVisible()
    })

    test('should render EditPage component with link to docFilePath', () => {
      render(<SubLessonPage {...props(false)} />)
      expect(
        screen.getByRole('link', { name: /edit this page/i })
      ).toHaveAttribute('href', expect.stringContaining(fakeGithubPath))
    })

    test('should match screenshot', async () => {
      const { container } = render(<SubLessonPage {...props(false)} />)

      expect(container).toMatchSnapshot()
    })

    test('should render correct lg breakpoint', () => {
      mockUseBreakpoint.mockReturnValueOnce(true)

      render(<SubLessonPage {...props(false)} />)

      expect(screen.getByTestId('sublesson__container')).toHaveStyle(
        'gridTemplateColumns: auto auto'
      )
      expect(screen.getByTestId('toc')).toHaveStyle('position: sticky')
    })

    test('should open accordion on mobile', async () => {
      mockUseBreakpoint.mockReturnValueOnce(false)

      render(<SubLessonPage {...props(false)} />)

      const toggle = screen.getByTestId('accordion-toggle')
      await userEvent.click(toggle)

      const accordionCollapse = screen.getByTestId('accordion-collapse')
      expect(accordionCollapse).toHaveClass('show')
    })
  })
})
