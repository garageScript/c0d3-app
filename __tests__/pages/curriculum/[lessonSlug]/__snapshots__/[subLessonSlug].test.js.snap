// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[subLessonSlug] Page should match screenshot 1`] = `
<div>
  <div
    class="mt-3 sublesson__container"
    data-testid="sublesson__container"
    style="grid-template-columns: none;"
  >
    <div
      class="card shadow-sm border-0 sublesson__sidebar"
      data-testid="toc"
      style="position: initial;"
    >
      <div
        class="accordion"
      >
        <div
          class="border-0 card"
        >
          <div
            class="py-3 toc__header card-header"
          >
            <button
              class="toc__toggle"
              data-testid="accordion-toggle"
              type="button"
            >
              Table of Contents
            </button>
          </div>
          <div
            class="accordion-collapse collapse"
            data-testid="accordion-collapse"
          >
            <div
              class="rounded toc__container card-body"
            >
              <li
                style="padding-left: 0rem;"
              >
                <a
                  class="MDX_a text-decoration-none"
                  href="#heading1"
                >
                  heading1
                </a>
              </li>
              <li
                style="padding-left: 1rem;"
              >
                <a
                  class="MDX_a text-decoration-none"
                  href="#heading2"
                >
                  heading2
                </a>
              </li>
              <li
                style="padding-left: 2rem;"
              >
                <a
                  class="MDX_a text-decoration-none"
                  href="#heading3"
                >
                  heading3
                </a>
              </li>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card shadow-sm d-flex flex-column gap-5 border-0 p-3 p-md-4 bg-white lesson-wrapper"
    >
      <div>
        <nav
          aria-label="Sub-lesson Links"
        >
          <a
            aria-current="page"
            class="subtitle link-dark dark d-block"
            href="/curriculum/js0/first_sub_lesson"
          >
            Part 1: first sub lesson
          </a>
          <a
            class="subtitle text-muted d-block"
            href="/curriculum/js0/second_sub_lesson"
          >
            Part 2: second sub lesson
          </a>
          <a
            class="subtitle text-muted d-block"
            href="/curriculum/js0/third_sub_lesson"
          >
            Part 3: third sub lesson
          </a>
        </nav>
        <h1
          class="MDX_h1"
          id="first"
        >
          first
        </h1>
        <p
          class="MDX_p"
        >
          simple
        </p>
        <div
          class="mt-5"
        >
          <div
            class="d-flex"
          >
            <a
              class="lessonLink next"
              href="/curriculum/js0/second_sub_lesson"
            >
              Next part: 
              <span>
                second sub lesson
              </span>
            </a>
          </div>
        </div>
      </div>
      <a
        class="MDX_a edit"
        href="https://github.com/garageScript/c0d3-app/edit/master/some/fake/path"
        title="Edit this page on Github"
      >
        Edit this page on Github
      </a>
    </div>
  </div>
</div>
`;
